impl/src/main/java/org/ehcache/config/builders/ResourcePoolsBuilder.java

```diff
@@ -69,25 +69,56 @@ public static ResourcePoolsBuilder newResourcePoolsBuilder() {
   public static ResourcePoolsBuilder newResourcePoolsBuilder(ResourcePools pools) {
     ResourcePoolsBuilder poolsBuilder = new ResourcePoolsBuilder();
     for (ResourceType currentResourceType : pools.getResourceTypeSet()) {
-      poolsBuilder = poolsBuilder.with(currentResourceType, pools.getPoolForResource(currentResourceType).getSize(),
-          pools.getPoolForResource(currentResourceType).getUnit(), pools.getPoolForResource(currentResourceType).isPersistent());
+      poolsBuilder = poolsBuilder.with(pools.getPoolForResource(currentResourceType));
     }
     return poolsBuilder;
   }
 
   /**
-   * Adds or replace the {@link ResourcePool} of {@link ResourceType} in the returned builder.
+   * Add the {@link ResourcePool} of {@link ResourceType} in the returned builder.
+   *
+   * @param resourcePool the non-{@code null} resource pool to add
+   * @return a new builder with the added pool
+   *
+   * @throws IllegalArgumentException if the set of resource pools already contains a pool for {@code type}
+   */
+  public ResourcePoolsBuilder with(ResourcePool resourcePool) {
+    final ResourceType type = resourcePool.getType();
+    final ResourcePool existingPool = resourcePools.get(type);
+    if (existingPool != null) {
+      throw new IllegalArgumentException("Can not add '" + resourcePool + "'; configuration already contains '" + existingPool + "'");
+    }
+    Map<ResourceType, ResourcePool> newPools = new HashMap<ResourceType, ResourcePool>(resourcePools);
+    newPools.put(type, resourcePool);
+    return new ResourcePoolsBuilder(newPools);
+  }
+
+  /**
+   * Add or replace the {@link ResourcePool} of {@link ResourceType} in the returned builder.
+   *
+   * @param resourcePool the non-{@code null} resource pool to add/replace
+   * @return a new builder with the added pool
+   */
+  // TODO: Determine the use case for this method
+  public ResourcePoolsBuilder withReplacing(ResourcePool resourcePool) {
```

Comment:
I was unable to locate a place where this method could be effectively used -- it may, at this point, be a candidate for removal.